🔨Refactor : 2차 리팩토링 완료

chabinhwang opened the pull request:
객체지향에 관한 책을 읽고 다시 한번 리팩토링 해보았습니다.

Commit:
submit

Commit:
🔨Refactor : 입력값 유효 확인 클래스 추가
Validation클래스를 통해 유효성을 체크

Commit:
🔨Refactor : 출력 전용 클래스 추가
PrintSequence로 결과값을 출력

Commit:
🔨Refactor : 기존 코드 수정
Validation과 PrintSequence를 활용하게 수정

Commit:
🔨Refactor : 주석 추가

Commit:
🔨Refactor : 테스트 케이스를 위해 전체적 수정-Lotto 클래스

Commit:
🔨Refactor : 테스트 케이스를 위해 전체적 수정-Validation 클래스

Commit:
🔨Refactor : 테스트 케이스를 위해 전체적 수정-PrintSequence 클래스

Commit:
🔨Refactor : 테스트 케이스를 위해 전체적 수정-MatchSequence 클래스

Commit:
🔨Refactor : 테스트 케이스를 위해 전체적 수정-InputSequence 클래스

Commit:
🔨Refactor : int 범위를 초과하는 입력에 대한 대응

Commit:
☔️Test: 테스트 코드 추가

Commit:
🔨Refactor:리팩토링 완료

chabinhwang committed:
🔨Refactor : 2차 리팩토링 완료

goldentrash commented on src/main/java/lotto/Application.java:
```diff
@@ -1,7 +1,35 @@
+        User user = new User();
+        // 초기 투자금 지정
+        user.setPurchaseMoney(io.inputInt());
+        System.out.println();
```
`IOSequence`라는 클래스를 구상하셨다면, `System.out`이 할 일도 모두 위임해보는건 어땠을까요? 만약 `io` 객체가 입력을 전담한다면 `inputSequence`와 같은 이름이 더 이해하기 쉬웠을 것 같습니다

goldentrash commented on src/main/java/lotto/Application.java:
```diff
@@ -1,7 +1,35 @@
+        int bonusNum = io.inputBonusInt(lt);
+        LottoGradeCheck lc=new LottoGradeCheck(lt, bonusNum);
+        // 등수 확인
+        user.checkLottoGrade(lc);
```
`user`가 `lc`의 결과를 직접 확인하네요. `user`의 책임에 `lc` 결과 확인을 포함시키신 이유가 있나요? `user`라는 이름을 통해 상상했던 것과는 조금 다른 일을 하고 있네요. 어쩌면 **`user`가 너무 많은 일을 하고 있지는 않을까요?**

chabinhwang replied:
`checkLottoGrade`를 어떤 회차(예:1240회차)의 당첨 번호 ( 주차가 다양하다면 checkLottoGrade객체가 다양하게 존재 )와 같이 지정한 다음
 "`user`가 **구매한 주차에 맞는 checkLottoGrade**를 활용해 번호를 대조해보며 등수를 확인하기" 를 의도하였습니다. `user`가 비교할 로또 당첨 회차가 바뀔 수 있다 라고 생각했습니다. 이때 checkLottoGrade의 메서드를 호출해 사용하는 방식을 채택하였습니다.

goldentrash replied:
문득 든 생각인데, 객체 구상에 있어 `User`를 현실 세계의 사람에 너무 과하게 투영한건 아닐까 우려되네요. 노파심에 덧붙이자면, 프로그램의 세계에서는 능동적으로 행동할 수 있는 주체가 사람이 유일하지 않다는 사실을 잊지 말아주세요!

chabinhwang replied:
흠...대대적으로 리팩토링하면서 든 생각인데 User가 하는 역할들이 Service에 해당하는 것 같아
**User가 없어도 될 것 같다** 는 생각이 들었습니다. 맞게 가고 있는지 모르겠네요..?

goldentrash commented on src/main/java/lotto/IOSequence.java:
```diff
@@ -0,0 +1,91 @@
+
+public class IOSequence {
+    private Scanner sc;
+    public IOSequence(Scanner sc) {
```
DI로군요! 좋습니다. 나중에 여유가 되시면 IoC Container를 구현해보시는것도 재밌을 거에요 :)

goldentrash commented on src/main/java/lotto/IOSequence.java:
```diff
@@ -0,0 +1,91 @@
+                input = Integer.parseInt(s);
+                if (input <= 0) {
+                    System.out.println("[ERROR]구입금액은 1 이상의 숫자를 입력해야 합니다.");
+                    continue;
```
`Early return`을 사용해 보는건 어땠을까요?

chabinhwang replied:
반복 조건을 제거함으로써 `예외_테스트()` 테스트 통과하도록 수정하였습니다!

goldentrash commented on src/main/java/lotto/IOSequence.java:
```diff
@@ -0,0 +1,91 @@
+                System.out.println("구입금액을 입력해 주세요.");
+                String s = sc.nextLine();
+                input = Integer.parseInt(s);
+                if (input <= 0) {
```
`IoSequence.inputInt()`의 역할이 `int32`를 입력받는게 아니라, **구입금액을 입력받는 것**일까요? 그렇다면 이 역할을 `IoSequence`가 책임지는게 맞을까요?

goldentrash commented on src/main/java/lotto/IOSequence.java:
```diff
@@ -0,0 +1,91 @@
+        }
+        return input;
+    }
+    public Lotto inputWinningLotto(){
```
만약, `Lotto`를 생성하는 **Business가 변경되면, 입출력 로직도 수정해야 할까요?** 물론 도메인 구상에 있어서, 변화하지 않는 내용을 어떻게 예상하는지에 따라서 Buisness의 변경이 Presentation가지 전파되는걸 피할 수 없기도 합니다. 

하지만 저는 `Lotto` 생성이 "숫자를 전부 입력받거나", "랜덤하게 생성하거나", "보너스 번호만 별도로 입력받거나" 하는 등 변화의 여지가 크다고 생각해요. `IoSequence`와 `Lotto`가 조금 더 독립적으로 동작할 수 있다면 좋을 것 같습니다 :)

goldentrash commented on src/main/java/lotto/LottoGradeCheck.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+public class LottoGradeCheck {
```
요구사항이 정리된 문서가 없었더라면 작성해주신 내용을 이해하기가 너무 어려웟을 것 같아요. `Enum`과 같이 상태를 **선언적**으로 표현해보는건 어땠을까요?

goldentrash commented on src/main/java/lotto/LottoPrize.java:
```diff
@@ -0,0 +1,19 @@
+package lotto;
+
+public enum LottoPrize {
+    FIRST(2_000_000_000.0),
```
`numeric separator` 좋네요~

goldentrash commented on src/main/java/lotto/Product/Lotto.java:
```diff
@@ -0,0 +1,60 @@
+            }
+            list.add(num);
+        }
+        list.sort(Integer::compareTo);
```
위 부분도 `private void validate(String numbers)`와 같이 메서드로 추출해보는건 어떨까요?

chabinhwang replied:
🔨`validate(String s)` 메서드를 추가 생성하여 생성자 간소화하였습니다!

goldentrash commented on src/main/java/lotto/Product/Lotto.java:
```diff
@@ -0,0 +1,60 @@
+        }
+    }
+
+    public List<Integer> getNumbers() {
```
State를 그대로 외부에 전달하는 메서드(흔히 `Getter`)는 유용합니다. 하지만 State를 가져다가 어딘가에 사용할 것이라면, `Lotto`가 스스로 그 일을 할 순 없을까요?

`LottoGradeCheck`의 코드 중 아래와 같은 내용을

```java
    private int matchBonusNumber(Lotto lotto){
        // 보너스번호가 로또 내 존재하면 1 아니면 0 반환
        if(lotto.getNumbers().contains(this.winningBonusNumber))
            return 1;
        return 0;
    }
```

다음과 같이 표현해보는건 어땠을까요?

```java
    private int matchBonusNumber(Lotto lotto){
        // 보너스번호가 로또 내 존재하면 1 아니면 0 반환
        if(lotto.matchBonusNumber(this.winningBonusNumber))
            return 1;
        return 0;
    }
```

chabinhwang replied:
LottoGradeCheck의 compareNumbers에서 Lotto의 numbers를 가져오는 대신 Lotto클래스 내부에서 진행하게 Refactor하였습니다!

goldentrash replied:
프로그램을 작성하실 때, `Getter`와 `Setter`가 너무 많다는 경보가 울린다면, 책임 경계가 명확한지 다시한번 점검해보시면 도움이 많이 되실거에요 :)

goldentrash commented on src/main/java/lotto/User.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.List;
+
+public class User {
+    private int purchaseMoney;
```
`prize`와 `purchaseMoney`와 같은 수치를 직접 설정해주고 있네요. 혹시 `static`하게 설정하지 않은 이유를 공유해주실 수 있을까요?

goldentrash commented on src/main/java/lotto/User.java:
```diff
@@ -0,0 +1,62 @@
+        this.gradeList=lottoGradeCheck.checkGrade(purchaseLottoList);
+    }
+
+    public void calculatePrize() {
```
`User`가 너무 다재다능한건 아닐까요?

goldentrash reviewed (commented):
고생하셨습니다! 부족하지만 리뷰 남겨보아요

PR을 올려주시기 전에 `bin`과 같이 생성되는 코드와 `DS_Store` 같은 환경 설정 파일은 제외해주시면 좋겠어요. 봐야 할 내용이 많아져서 혹시 중요한 부분을 놓칠수도 있으니까요

주석을 열심히 달아주셔서 읽기 편했습니다. 하지만 가독성을 향상시키기 위해서는 주석보다도 중요한 부분이 많습니다. 메서드와 메서드, 혹은 로직과 로직 사이의 **적절한 줄띄움**을 통해 로직을 그룹화하는건 읽는데 큰도움이 돼요! 그렇게 모아진 부분이 충분히 복잡해진다면 별도 메서드로 추출하기도 쉬워지겠죠?

더해서, 좋은 주석 이상으로 **좋은 이름**도 중요합니다. `lc`를 `LottoCheckGrade`의 줄임말로 사용하신듯 한데, 조금 길어지더라도 `lottoGrade`와 같은 이름이었다면 코드를 이해하기가 훨씬 수월햇을 것 같아요 :)

마지막으로, Spring을 공부하시면서 자연스럽게 Layered Architecture를사용하고 계실거같아요. 혹시 Application(혹은 Business) Layer를 Domain과 Service로 구분하는 이유에 대해 고민해보신적 있으실까요? 둘 모두 비즈니스 로직을 담당한다면 어째서 둘을 나누는게 일반적일까요? 둘의 역할 차이를 고민해보신다면 도움이 많이 될 거라고 믿습니다!

chabinhwang committed:
🔨Refactor :
1. LottoGradeCheck의 compareNumbers에서 Lotto의 numbers를 가져오는 대신 Lotto클래스 내부에서 진행
2. IOSequence에서 inputInt에 while문 제거를 통해 예외_테스트 통과

chabinhwang committed:
🔨Refactor : 추가 메서드 생성으로 Lotto 생성자 간소화

chabinhwang committed:
🔨Refactor :
1. 버퍼 문제로 Application에 존재하던 print문 제거
2. IOSequence 메서드명 수정

goldentrash reviewed (commented):
카카오톡으로 말씀드렸던것처럼, 우선은 Source 리뷰보다는 객체 관계 설계를 먼저 리뷰해보도록 할게요.

TIP. `Files Changed`에서 코멘트를 작성하신 후, `Start Review` 버튼을 클릭하시면 코멘트를 펜딩하셨다가 한 번에 달아주실 수 있어요! 
(개별 코멘트는 모두 별도 알람 수신으로 연결되어서 혹여나 저처럼 알람을 켜둔 사람은 알람 폭탄을 받을 수 있습니다 ㅠ)

chabinhwang committed:
🔨Refactor : Config를 통한 의존성 주입. Controller를 통한 게임 진행.
